kratos/utilities/python_function_callback_utility.h

```diff
@@ -8,214 +8,357 @@
+     * @param LocalSystem The parameters defining the local system
+     */
+    PythonGenericFunctionUtility(
+        const std::string& rFunctionBody,
+        Parameters LocalSystem = Parameters{}
+        ) : mFunctionBody(rFunctionBody)
     {
-        //compile the function starting from the string function body
-        try
-        {
-            main_module = pybind11::module::import("__main__");
-            main_namespace = main_module.attr("__dict__");
-            pybind11::exec("from math import *", main_namespace);
-            mfunction_body = function_body;
-            //mbytecode = pybind11::object( pybind11::handle<>( (PyObject*) Py_CompileString(function_body.c_str(), "pyscript", Py_eval_input) ) );
-        }
-        catch(pybind11::error_already_set const&)
-        {
+        // Compile the function starting from the string function body
+        try {
+            mMainModule = pybind11::module::import("__main__");
+            mMainNameSpace = mMainModule.attr("__dict__");
+            pybind11::exec("from math import *", mMainNameSpace);
+//             mByteCode = pybind11::object( pybind11::handle<>( (PyObject*) Py_CompileString(rFunctionBody.c_str(), "pyscript", Py_eval_input) ) );
+        } catch(pybind11::error_already_set const&) {
             PyErr_Print();
         }
 
+        // Here get the local system if it is provided
+        if(LocalSystem.Has("origin")) {
+            mUseLocalSystem = true;
 
-        // here get the local system if it is provided
-        if(local_system.Has("origin"))
-        {
-            muse_local_system = true;
-
-            for(unsigned int i = 0; i<3; ++i)
-            {
-                mxc[i] = local_system["origin"][i].GetDouble();
-                for(unsigned int j = 0; j<3; ++j)
-                {
-                    mR(i, j) = local_system["axes"][i][j].GetDouble();
+            for(IndexType i = 0; i<3; ++i) {
+                mCenterCoordinates[i] = LocalSystem["origin"][i].GetDouble();
+                for(IndexType j = 0; j<3; ++j) {
+                    mRotationMatrix(i, j) = LocalSystem["axes"][i][j].GetDouble();
                 }
             }
-        }
-        else
-        {
-            muse_local_system = false;
+        } else {
+            mUseLocalSystem = false;
         }
 
-        // check if it depends on space
-        if (function_body.find(std::string("x"))==std::string::npos &&
-                function_body.find(std::string("y")) ==std::string::npos &&
-                function_body.find(std::string("z")) ==std::string::npos)
-        {
-            mdepends_on_space = false;
+        // Check if it depends on space
+        if (rFunctionBody.find(std::string("x")) == std::string::npos &&
+            rFunctionBody.find(std::string("y")) == std::string::npos &&
+            rFunctionBody.find(std::string("z")) == std::string::npos &&
+            rFunctionBody.find(std::string("X")) == std::string::npos &&
```

Comment:
`X` or `x0`
I don't have a strong preference, I would be ok with the current solution (is more consistent with literature)

Reply:
And the current solution is the only one will work. When python looks for x and there is a x0 an error will be thrown as the 0 does not have any operation. On the other hand as python is case sensitive X does not give any issue

Reply:
then it is decided :)